odoo/addons/base/models/ir_qweb.py

```diff
@@ -105,25 +547,157 @@ def _render(self, template, values=None, **options):
+
+            render(qweb, values)
+
+        where ``qweb`` is a QWeb instance and ``values`` are the values to render.
+        """
+        if options is None:
+            options = {}
+
+        element, document, ref = self._get_template(template, options)
+        if not ref:
+            ref = element.get('t-name', str(document))
+
+        element.attrib.pop('t-name', None)
+
+        # reference to get xml and etree (usually the template ID)
+        options['ref'] = ref
+        # str xml of the reference template used for compilation. Useful for debugging, dev mode and profiling.
+        options['ref_xml'] = document if isinstance(document, str) else str(document, 'utf-8')
+
+        _options = dict(options)
+        options = frozendict(options)
+
+        # Initial template value send to render method (not in the froozen dict because it may be
+        # different from one render to another. Indeed, it may be the view ID or the key)
+        _options['template'] = template
+        # Root of the etree which will be processed during compilation.
+        _options['root'] = element.getroottree()
+        # Reference to the last node being compiled. It is mainly used for debugging and displaying
+        # error messages.
+        _options['last_path_node'] = None
+
+        if not options.get('nsmap'):
+            _options['nsmap'] = {}
+
+        # generate code
+
+        def_name = f"template_{ref}" if isinstance(ref, int) else "template"
+
+        try:
+            _options['_text_concat'] = []
+            self._append_text("", _options) # To ensure the template function is a generator and doesn't become a regular function
+            code_lines = ([dedent(f"""
+                def {def_name}(self, values, log):
+                    attrs = None
+                """)] +
+                self._compile_node(element, _options, 1) +
+                self._flush_text(_options, 1))
+        except QWebException as e:
+            raise
+        except Exception as e:
+            raise QWebException("Error when compiling xml template",
+                self, options, template, path=_options.get('last_path_node')) from e
+        try:
+            code = '\n'.join(code_lines)
+        except Exception as e:
+            code = '\n'.join(map(str, code_lines))
+            raise QWebException("Error when compiling xml template",
+                self, options, template, code=code) from e
+
+        # compile code and defined default values
+
         try:
-            id_or_xml_id = int(id_or_xml_id)
+            # noinspection PyBroadException
+            compiled = compile(code, f'<{def_name}>', 'exec')
+            globals_dict = self._prepare_globals({}, options)
+            globals_dict['__builtins__'] = globals_dict # So that unknown/unsafe builtins are never added.
+            unsafe_eval(compiled, globals_dict)
+            compiled_fn = globals_dict[def_name]
+        except QWebException as e:
```

Comment:
Just a simple linting issue: `Unused variable 'e', pylint(unused-variable)`. (not the only one)